\documentclass{article}

% if you need to pass options to natbib, use, e.g.:
% \PassOptionsToPackage{numbers, compress}{natbib}
% before loading nips_2017
%
% to avoid loading the natbib package, add option nonatbib:
% \usepackage[nonatbib]{nips_2017}

\usepackage{nips_2017}

% to compile a camera-ready version, add the [final] option, e.g.:
% \usepackage[final]{nips_2017}

\usepackage[utf8]{inputenc} % allow utf-8 input
\usepackage[T1]{fontenc}    % use 8-bit T1 fonts
\usepackage{hyperref}       % hyperlinks
\usepackage{url}            % simple URL typesetting
\usepackage{booktabs}       % professional-quality tables
\usepackage{amsfonts}       % blackboard math symbols
\usepackage{nicefrac}       % compact symbols for 1/2, etc.
\usepackage{microtype}      % microtypography
\usepackage{graphicx}
\usepackage{pgfplotstable}
\usepackage{float}
\usepackage{csvsimple}
\usepackage[normalem]{ulem}
\graphicspath{{/Users/fonz/Documents/Projects/GTEx/plotting/}}

\title{Explaining gene expression variation to neural network derived image features}

% The \author macro works with any number of authors. There are two
% commands used to separate the names and addresses of multiple
% authors: \And and \AND.
%
% Using \And between authors leaves it to LaTeX to determine where to
% break the lines. Using \AND forces a line break at that point. So,
% if LaTeX puts 3 of 4 authors names on the first line, and the last
% on the second line, try using \AND instead of \And before the third
% author name.

\author{
  William Jones\thanks{Use footnote for providing further
    information about author (webpage, alternative
    address)---\emph{not} for acknowledging funding agencies.} \\
  Department of Computer Science\\
  Cranberry-Lemon University\\
  Pittsburgh, PA 15213 \\
  \texttt{hippo@cs.cranberry-lemon.edu  asdfas} \\
  %% examples of more authors
  %% \And
  %% Coauthor \\
  %% Affiliation \\
  %% Address \\
  %% \texttt{email} \\
  %% \AND
  %% Coauthor \\
  %% Affiliation \\
  %% Address \\
  %% \texttt{email} \\
  %% \And
  %% Coauthor \\
  %% Affiliation \\
  %% Address \\
  %% \texttt{email} \\
  %% \And
  %% Coauthor \\
  %% Affiliation \\
  %% Address \\
  %% \texttt{email} \\
}

\begin{document}
% \nipsfinalcopy is no longer used

\maketitle

\begin{abstract}
How does gene expression influence visual characteristic of tissues? Can the variation in expression be explaining using data extracting from these images alone? To what extent is this possible? In this work, I am to answer these questions, by employing neural networks on a dataset of biomedical images annotated with genotype and gene expression data.
\end{abstract}

\section{Introduction}
Biomedical images are routinely used by doctors in pathology to diagnose diseases. For example, lung biopsies are taken from patients when grading different levels of lung cancers. Specifically for lung cancer, a great deal of work has focussed on the identification of specific tissue characteristics which are indicative of metastasis. Indeed, this has been the topic of many supervised machine learning competition and approaches. We can now accurately identify metastatic regions from high resolution histopathology lung images, with the state of the art being achieve by Convolutional Neural Networks (CNNs).

What other types of information can be estimated from these high resolution images? We explore this direction using a datasets of.....


\section{Results}
We find that across 10 different tissue types, large amount of variation in expression can be explained in large part with a combination of technical factors and image features. Gene expression reflects the biological activity of a cell at a given point in time. Technical factors are already known to have a large impact on gene expression. For example, Ischemic time (SMTISCH), reflects the period time between the donor's death, and tissue extraction. During apoptosis, multiple biological pathways are activated which have a profound impact on the biological activity within a cell. This is reflected by the fact that Ischemic time explains the largest amount of variance in expression out of all of the technical factors. With respect to the image features using the described method, more variance is explained when using large patch-sizes. This is surprising because intuitively one might imagine that variation in gene expression might affect only characteristics in tissues visible at smaller scales. However, these results provide evidence against this intuition.

\section{Future work}
This study motivates further work that aims to close the information boundary between biomedical images and gene expression. Future work will aim to predict variation in other forms of biomarkers, for example expression of particular genes that might be biomarkers for a given diagnosis. This work will require more comprehensive datasets, but with modern day neural networks able to successfully perform function approximation for a wide variety of complex tasks, this might soon become a reality. 


\section*{Plan}

\subsection*{Q1: How much variability is shared between image features and expression?} 
\begin{enumerate}
\item \sout{Calculate PCA of image features and expression}

\item \sout{Calculate cross correlations between the PCs until 99.9\% variance is explained.}

\item \sout{Use the cross correlation measurements to derive total fraction of variance explained of images by expression: calculate $\sum_i \lambda_i \sum_j r^2_{ij} $ where $\lambda_i$ is the $i$th eigenvalue of the image feature PCA, $r^2_{ij}$ is the correlation of image feature PC $i$ with expression feature PC $j$.}

\item \sout{Make all of the above one-line-to-run for any combination of patch size, feature layer, tissue}

\item \sout{Report the value for each patch size, feature layer, tissue;} interpret findings

\end{enumerate}

\subsection*{Q2: How much of the shared variability is due to technical confounders?} 

\begin{enumerate}
\item \sout{Regress out known confounders $C$ from the image feature $F$ and expression data E. I.e., fit models $F \sim C + \epsilon$ and $E \sim C + \epsilon$. Use log scale for Ischemic time, monitor all scatter plots for confounders that are included in the model that the linear model assumptions hold. Decide whether to retain full model of all confounders x all genes/features, retain only significant associations, or do forward feature selection.}

\item \sout{Repeat steps Q1:1-5 above using the residuals of this model. }

\item \sout{Create a scatter plot of total shared variance vs. technical shared variance [again, all tissues etc.]}

\item Interpret findings in context of many tissues, scales, feature layers
\end{enumerate}

\subsection*{Q3: What genes share expression variability to visual features, and at which scales?}
\begin{enumerate}
\item \sout{Using residuals of technical confounders, fit linear model of expression vs. image feature - calculate effect sizes and p-values}

\item \sout{Apply multiple testing correction to derive a list of image feature associated genes.}

\item \sout{Sort genes based on variance explained by each image feature, perform GSEA or gProfiler ranked gene list analysis to test whether particular aspects of biology are enriched.} Interpret

\item Assess number of associations [total and per-feature] for different tissues, feature scales. Assess overlap of findings from gene list analyses. Interpret.
\end{enumerate}

\subsection*{Q3A: What genes are influenced by technical confounders?}
\begin{enumerate}

\item For each confounder, calculate the fraction of variance it explains for each gene expression level. 

\item Sort genes based on variance explained by confounder, perform GSEA or gProfiler ranked gene list analysis to test whether particular aspects of biology are enriched.]

\end{enumerate}

\subsection*{Once these done}
\begin{itemize}
\item Q4: What do the image features represent?
\item Q5: Is there a genetic basis to the image features?
\item Q6: Are there gene expression levels that cause image feature changes, or vice versa?
\item Q7: Are there image features that are associated with clinical annotations? Are any causal?

\end{itemize}

\section*{Methods}

\subsection*{Assessing shared variability between image features and expression}
We calculate principle components that capture 99.9\% of variation in the image features and expression features respectively. We repeat this after fitting a linear model containing known technical confounders to both the image features and expression, and regressing out these effects.

We choose the top 2000 transcripts with mean expression greater than one, and look for associations with the image features across tissue samples. For image features generated from Lung tissue, we plot a QQ-plot with 

\section*{Results}

Variability in expression explains a substantial amount of variability in the image features (Figure ???). Furthermore, almost 50\% of the variability in expression, and almost 40\% of the variability in the image features, is explained by 51 known technical factors (Figure ???)

\begin{figure}[H]
  \centering
    \includegraphics[width=0.5\linewidth]{/NIPS/compare_variance_explained} 
  \caption{Sample figure caption.}
\end{figure}


\begin{figure}[H]
  \centering
    \includegraphics[width=1\linewidth]{/TFCorrectedFeatureAssociations/tf_feature_selection_expression} 
  \caption{Sample figure caption.}
\end{figure}

\begin{figure}[H]
  \centering
    \includegraphics[width=1\linewidth]{/TFCorrectedFeatureAssociations/tf_feature_selection_image_features} 
  \caption{Sample figure caption.}
\end{figure}

%\pgfplotstableset{col sep=comma}{plotting/NIPS/ontology_results.csv}


\begin{table}[H]
\caption{Enriched Ontology terms for image feature 501}
\label{tab:image_feature_pcs_and_tfs}       % Give a unique label

\begin{tabular}{p{2cm}p{6cm}p{6cm}}
\hline\noalign{\smallskip}
p-value & Ontology & Genes  \\
1.21e-15 & lamellar body & LAMP3,CTSH,SFTPA1,NAPSA,SFTPD  \\
6.87e-09 & Defective ABCA3 causes pulmonary surfactant metabolism dysfunction type 3 (SMDP3) & SFTPA1,SFTPD,SFTPB,SFTPA2 \\


%\noalign{\smallskip}\svhline\noalign{\smallskip}
\noalign{\smallskip}\hline\noalign{\smallskip}
\end{tabular}
% $^a$ Table foot note (with superscript)
\end{table}

\begin{table}[H]
\caption{Enriched Ontology terms for image feature 3}
\label{tab:image_feature_pcs_and_tfs}       % Give a unique label

\begin{tabular}{p{2cm}p{6cm}p{6cm}}
\hline\noalign{\smallskip}
p-value & Ontology & Genes  \\
0.0498 & BRONCHIECTASIS WITH OR WITHOUT ELEVATED SWEAT CHLORIDE 3; BESC3;;CYSTIC FIBROSIS-LIKE SYNDROME & SCNN1G \\


%\noalign{\smallskip}\svhline\noalign{\smallskip}
\noalign{\smallskip}\hline\noalign{\smallskip}
\end{tabular}
% $^a$ Table foot note (with superscript)
\end{table}








%[1] Alexander, J.A.\ \& Mozer, M.C.\ (1995) Template-based algorithms
%for connectionist rule extraction. In G.\ Tesauro, D.S.\ Touretzky and
%T.K.\ Leen (eds.), {\it Advances in Neural Information Processing
%  Systems 7}, pp.\ 609--616. Cambridge, MA: MIT Press.
%
%[2] Bower, J.M.\ \& Beeman, D.\ (1995) {\it The Book of GENESIS:
%  Exploring Realistic Neural Models with the GEneral NEural SImulation
%  System.}  New York: TELOS/Springer--Verlag.
%
%[3] Hasselmo, M.E., Schnell, E.\ \& Barkai, E.\ (1995) Dynamics of
%learning and recall at excitatory recurrent synapses and cholinergic
%modulation in rat hippocampal region CA3. {\it Journal of
%  Neuroscience} {\bf 15}(7):5249-5262.




\end{document}
