%%%%%%%%%%%%%%%%%%%%%%%% referenc.tex %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% sample references
% %
% Use this file as a template for your own input.
%
%%%%%%%%%%%%%%%%%%%%%%%% Springer-Verlag %%%%%%%%%%%%%%%%%%%%%%%%%%
%


\begin{thebibliography}{99.}

\bibitem{imaging-as-a-tool}
Frangou, S., \& Murray, R. M. (1996). Imaging as a tool in exploring the neurodevelopment and genetics of schizophrenia. British Medical Bulletin, 52(3), 587?596. http://doi.org/10.1093/oxfordjournals.bmb.a011569

\bibitem{imaging-genetics} MUÑOZ KE, HYDE LW, HARIRI AR. Imaging Genetics. Journal of the American Academy of Child and Adolescent Psychiatry. 2009;48(4):356-361. doi:10.1097/CHI.0b013e31819aad07.

\bibitem{imaging-genetics-parkinsons} Kim, M., Kim, J., Lee, S.-H., \& Park, H. (2017). Imaging genetics approach to Parkinson?s disease and its correlation with clinical score. Scientific Reports, 7, srep46700. http://doi.org/10.1038/srep46700

\bibitem{hematoxylin-and-eosin-staining} Fischer, A. H., Jacobson, K. A., Rose, J., \& Zeller, R. (2008). Hematoxylin and Eosin Staining of Tissue and Cell Sections. Cold Spring Harbor Protocols, 2008(5), pdb.prot4986?pdb.prot4986. http://doi.org/10.1101/pdb.prot4986

\bibitem{histopathological-image-analysis} Gurcan MN, Boucheron L, Can A, Madabhushi A, Rajpoot N, Yener B. Histopathological Image Analysis: A Review. IEEE reviews in biomedical engineering. 2009;2:147-171. doi:10.1109/RBME.2009.2034865.

\bibitem{GTEx-project} Lonsdale, J., Thomas, J., Salvatore, M., Phillips, R., Lo, E., Shad, S., et al. (2013). The Genotype-Tissue Expression (GTEx) project. Nature Genetics, 45(6), 580?585. http://doi.org/10.1038/ng.2653

\bibitem{GTEx-histology} GTEx Image Viewer https://brd.nci.nih.gov/brd/image-search/searchhome

\bibitem{what-is-complex-about-complex-disorders} Mitchell, K. J. (2012). What is complex about complex disorders? Genome Biology, 13(1), 237. http://doi.org/10.1186/gb-2012-13-1-237

\bibitem{gene-expression-parkinsons} Lewis, P. A., \& Cookson, M. R. (2012). Gene expression in the Parkinson's disease brain. Brain Research Bulletin, 88(4), 302?312. http://doi.org/10.1016/j.brainresbull.2011.11.016

\bibitem{complex-sources-of-variation} McCall, M. N., Illei, P. B., \& Halushka, M. K. (2016). Complex Sources of Variation in Tissue Expression Data: Analysis of the GTEx Lung Transcriptome. The American Journal of Human Genetics, 99(3), 624€“635. http://doi.org/10.1016/j.ajhg.2016.07.007

\bibitem{image-net} Krizhevsky, A., Sutskever, I., \& Hinton, G. E. (2012). ImageNet Classification with Deep Convolutional Neural Networks. Advances in Neural, 1097â€“1105.

\bibitem{scalable-unsupervised-learning} Lee, H., Grosse, R., Ranganath, R., \& Ng, A. Y. (2009). Convolutional deep belief networks for scalable unsupervised learning of hierarchical representations. Proceedings of the 26th annual (pp. 609€“616). ACM. http://doi.org/10.1145/1553374.1553453

\bibitem{deep-inside-convolutional-networks} Simonyan, K., Vedaldi, A., \& Zisserman, A. (2013, December 20). Deep Inside Convolutional Networks: Visualising Image Classification Models and Saliency Maps. arXiv.org.

\bibitem{fine-tuning-deep-convolutional-neural-networks} Gando, G., Yamada, T., Sato, H., Oyama, S., \& Kurihara, M. (2016). Fine-tuning deep convolutional neural networks for distinguishing illustrations from photographs. Expert Systems with Applications, 66, 295â€“301. http://doi.org/10.1016/j.eswa.2016.08.057

\bibitem{network-in-network} Lin, M., Chen, Q., \& Yan, S. (2013, December 16). Network In Network. arXiv.org.

\bibitem{autoencoders-unsupervised-learning-and-deep-architectures} Baldi, P. (2012). Autoencoders, unsupervised learning, and deep architectures. Presented at the Proceedings of ICML Workshop on Unsupervised and ?.

\bibitem{shapiro-computer-vision} Shapiro, L. G. \& Stockman, G. C: "Computer Vision", page 137, 150. Prentice Hall, 2001

\bibitem{otsu-method} Otsu, N. (1979). A Threshold Selection Method from Gray-Level Histograms. IEEE Transactions on Systems, Man, and Cybernetics, 9(1), 62?66. http://doi.org/10.1109/TSMC.1979.4310076

\bibitem{keras} Keras Chollet, Fran\c{c}ois and others, https://github.com/fchollet/keras

\bibitem{openslide} Goode. (2012). OpenSlide: A vendor-neutral software foundation for digital pathology. Journal of Pathology Informatics, 4(1), 27. http://doi.org/10.4103/2153-3539.119005

\bibitem{generative-adverserial-networks} Goodfellow, I., Pouget-Abadie, J., Mirza, M., Xu, B., Warde-Farley, D., Ozair, S., et al. (2014). Generative Adversarial Nets. Advances in Neural ?, 2672?2680.

\bibitem{gans-biological-image-synthesis} Osokin, A., Chessel, A., Salas, R. E. C., \& Vaggi, F. (2017, August 15). GANs for Biological Image Synthesis. arXiv.org.

\bibitem{variational-autoencoders} Kingma, D. P., \& Welling, M. (2013, December 20). Auto-Encoding Variational Bayes. arXiv.org.

\bibitem{fully-convolutional-networks} Long, J., Shelhamer, E., \& Darrell, T. (2015). Fully Convolutional Networks for Semantic Segmentation. Proceedings of the IEEE ?, 3431?3440.

\bibitem{sparse-autoencoders} Hou, L., Nguyen, V., Samaras, D., Kurc, T. M., Gao, Y., Zhao, T., \& Saltz, J. H. (2017, April 3). Sparse Autoencoder for Unsupervised Nucleus Detection and Representation in Histopathology Images. arXiv.org.

\bibitem{segmenting-classifying-epithelial} Xu, J., Luo, X., Wang, G., Gilmore, H., \& Madabhushi, A. (2016). A Deep Convolutional Neural Network for segmenting and classifying epithelial and stromal regions in histopathological images. Neurocomputing, 191, 214?223. http://doi.org/10.1016/j.neucom.2016.01.034

\bibitem{mendelian-randomization} Smith, G. D., \& Ebrahim, S. (2008). Mendelian Randomization: Genetic Variants as Instruments for Strengthening Causal Inference in Observational Studies.

\bibitem{hipsci-project} Streeter, I., Harrison, P. W., Faulconbridge, A., Flicek, P., Parkinson, H., \& Clarke, L. (2017). The human-induced pluripotent stem cell initiative?data resources for cellular genetics. Nucleic Acids Research, 45(D1), D691?D697. http://doi.org/10.1093/nar/gkw928

\end{thebibliography}



